app/scripts/lib/hyperliquid-referral-middleware.ts

```diff
@@ -0,0 +1,87 @@
+import {
+  createAsyncMiddleware,
+  type AsyncJsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import { ValidPermission, type Caveat } from '@metamask/permission-controller';
+import type {
+  Json,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+import log from 'loglevel';
+
+export const HYPERLIQUID_ORIGIN = 'https://app.hyperliquid.xyz';
+
+export type ExtendedJSONRPCRequest = JsonRpcRequest & {
+  origin: string;
+  tabId: number;
+};
+
+export enum HyperliquidPermissionTriggerType {
+  NewConnection = 'new_connection',
+  OnNavigateConnectedTab = 'on_navigate_connected_tab',
+}
+
+function isExtendedJSONRPCRequest(req: JsonRpcRequest): req is ExtendedJSONRPCRequest {
+  return (
+    Boolean((req as ExtendedJSONRPCRequest).origin) &&
+    Boolean((req as ExtendedJSONRPCRequest).tabId)
+  );
+}
+
+/**
+ * Creates middleware that monitors permission requests for Hyperliquid.
+ * When a permission is granted to Hyperliquid, it triggers the referral flow.
+ *
+ * @param handleHyperliquidReferral - Function to handle the referral flow
+ * @returns Middleware function
+ */
+export function createHyperliquidReferralMiddleware(
+  handleHyperliquidReferral: (args: {
+    origin: string;
+    tabId: number;
+    triggerType: HyperliquidPermissionTriggerType;
+  }) => void,
+) {
+  return createAsyncMiddleware(
+    async (
+      req: JsonRpcRequest,
+      res: PendingJsonRpcResponse<
+        ValidPermission<string, Caveat<string, Json>>[]
+      >,
+      next: AsyncJsonRpcEngineNextCallback,
+    ) => {
+      if (!isExtendedJSONRPCRequest(req)) {
+        return;
+      }
+
+      // First, call next to process the request
+      await next();
+
+      // After the request is processed, check if it was a successful permission grant for Hyperliquid
+      const isHyperliquidConnectionRequest =
+        req.method === 'wallet_requestPermissions' &&
```

Comment:
technically need to handle eth_requestAccounts here as well

Reply:
would not be work if hyperliquid moved to using the multichain api (`wallet_createSession`) directly or indirectly, but I don't think we need to account for that right now, just a btw